File: src/Simbuino.Emulator/ADC.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Simbuino.Emulator
{
	// this class emulates the AVRs ADC circuitry
	public class ADC
	{
		public static ObservableRegister ADSCRA;
		public static ObservableRegister ADMUX;
		public static ObservableRegister ADCH;
		public static ObservableRegister ADCL;

		// hard-coded for now...
		private int BatteryLevel = 0x308;
		private int AmbientLightSensor = 0x00;

		public ADC()
		{
			ADSCRA = AtmelContext.RAM[AtmelIO.ADSCRA] as ObservableRegister;
			ADSCRA.OnRegisterChanged += ADSCRA_OnRegisterChanged;
			ADMUX = AtmelContext.RAM[AtmelIO.ADMUX] as ObservableRegister;
			ADCH = AtmelContext.RAM[AtmelIO.ADCH] as ObservableRegister;
			ADCH.OnRegisterRead += ADCH_OnRegisterRead;
			ADCL = AtmelContext.RAM[AtmelIO.ADCL] as ObservableRegister;
			ADCL.OnRegisterRead += ADCL_OnRegisterRead;
		}

		void ADSCRA_OnRegisterChanged(int oldVal, int newVal)
		{
			// finish adc conversions immediately
			if ((newVal & (1 << AtmelIO.ADSC)) != 0)
				ADSCRA.Value = newVal ^ (1 << AtmelIO.ADSC);
		}

		void ADCH_OnRegisterRead(ref int val)
		{
			// get the ADC channel num
			int adc = (ADMUX.Value & 0x0f);

			// fetch the appropriate value
			int result =
				(adc == 6) ? this.BatteryLevel :
				(adc == 7) ? this.AmbientLightSensor :
				0;

			// the ADLAR bit determines the format
			if (ADMUX[AtmelIO.ADLAR] == 0)
				val = (result >> 8) & 0x03;
			else
				val = (result >> 2) & 0xff;
		}

		void ADCL_OnRegisterRead(ref int val)
		{
			// get the ADC channel num
			int adc = (ADMUX.Value & 0x0f);

			// fetch the appropriate value
			int result =
				(adc == 6) ? this.BatteryLevel :
				(adc == 7) ? this.AmbientLightSensor :
				0;

			// the ADLAR bit determines the format
			if (ADMUX[AtmelIO.ADLAR] == 0)
				val = result & 0xff;
			else
				val = (result << 6) & 0xC0;
		}
		

	}
}

File: binaries/Simbuino4Web/Scripts/Views/ADC.js
$(function () {

	// this class emulates the AVRs ADC circuitry
	ADC = {
		BatteryLevel: 0x308,
		AmbientLightSensor: 0x00,

		Init: function()
		{
			var self = this;
			this.ADSCRA = AtmelContext.RAM[AtmelIO.ADSCRA];
			this.ADSCRA.OnRegisterChanged.push(function (oldVal, newVal) { self.ADSCRA_OnRegisterChanged(oldVal, newVal); });
			this.ADMUX = AtmelContext.RAM[AtmelIO.ADMUX];
			this.ADCH = AtmelContext.RAM[AtmelIO.ADCH];
			this.ADCH.OnRegisterRead.push(function (oldVal, newVal) { self.ADCH_OnRegisterRead(oldVal, newVal); });
			this.ADCL = AtmelContext.RAM[AtmelIO.ADCL];
			this.ADCL.OnRegisterRead.push(function (oldVal, newVal) { self.ADCL_OnRegisterRead(oldVal, newVal); });
		},

		ADSCRA_OnRegisterChanged: function(oldVal, newVal)
		{
			// finish adc conversions immediately
			if ((newVal & (1 << AtmelIO.ADSC)) != 0)
				this.ADSCRA.set(newVal ^ (1 << AtmelIO.ADSC));
		},

		ADCH_OnRegisterRead: function(val)
		{
			// get the ADC channel num
			var adc = (this.ADMUX.get() & 0x0f);

			// fetch the appropriate value
			var result =
			(adc == 6) ? this.BatteryLevel :
			(adc == 7) ? this.AmbientLightSensor :
			0;


			this.ADCL = AtmelContext.RAM[AtmelIO.ADCL];
			this.ADCL.OnRegisterRead.push(function (oldVal, newVal) { self.ADCL_OnRegisterRead(oldVal, newVal); });
		},

		ADSCRA_OnRegisterChanged: function(oldVal, newVal)
		{
			// finish adc conversions immediately
			if ((newVal & (1 << AtmelIO.ADSC)) != 0)
				this.ADSCRA.set(newVal ^ (1 << AtmelIO.ADSC));
		},

		ADCH_OnRegisterRead: function(val)
		{
			// get the ADC channel num
			var adc = (this.ADMUX.get() & 0x0f);

			// fetch the appropriate value
			var result =
			(adc == 6) ? this.BatteryLevel :
			(adc == 7) ? this.AmbientLightSensor :
			0;

			// the ADLAR bit determines the format
			if (this.ADMUX.get_bit(AtmelIO.ADLAR) == 0)
				return (result >> 8) & 0x03;
			else
				return (result >> 2) & 0xff;
		},

		ADCL_OnRegisterRead: function(val)
		{
			// get the ADC channel num
			var adc = (this.ADMUX.get() & 0x0f);

			// fetch the appropriate value
			var result =
			(adc == 6) ? this.BatteryLevel :
			(adc == 7) ? this.AmbientLightSensor :
			0;

			// the ADLAR bit determines the format
			if (this.ADMUX.get_bit(AtmelIO.ADLAR) == 0)
				return result & 0xff;
			else
				return (result << 6) & 0xC0;
		}

	}

});
